torch/distributed/pipelining/schedules.py

```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
         stage_index_to_stage: Dict[int, _PipelineStageBase] = {
             stage.stage_index: stage for stage in self._stages
         }
-        prev_rank: int = (self.rank - 1) % self.pp_group_size
-        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        # determine prev_rank and next_rank based on which ranks are next to
+        # the stages in the pipeline_order
+        all_prev_ranks: Set[int] = set()
+        all_next_ranks: Set[int] = set()
+        for stage_index in stage_index_to_stage.keys():
```

Comment:
is it safe to loop over stages here?

R0 [S0, S7]
R1 [S1, S6]
R2 [S2, S5]
R3 [S3, S4]

For rank 2- what if S2 needs to send forward to S3 for a later mb_chunk and S5 needs to recv forward from S4 for an earlier chunk?  This is just one attempt to construct an example where 2 pairs of stages are communicating across rank boundary at the same timestep.

Do we need to batch comms for S2-S3 with comms for S4-S5?

IIUC the current for loop separates batching to stage-boundary.  I think batching only needs to be separated by rank boundary, and may be required to be batched across stage-boundary in some cases.  Wdyt?

Reply:
ok, i'm wrong, sorry. I see that you are building a list of ops and then later executing the whole list. So you are already batching across stages, it doesn't really matter about the for loop.

Reply:
In the current runtime, there can only be a single action per timestep, which means that only 1 stage / 1 mb per rank can operate at any point in time.

So the case with S2 -> S3 and S5 -> S4 wouldn't happen in the same timestep. This change doesn't break that assumption it just changes the number of ranks that any rank can communicate with.

Previously a rank could only communicate with 0-2 other ranks (forward/backward), now a rank can potentially communicate with all ranks.